src/Router.sol

```diff
@@ -170,28 +172,61 @@ contract Router is ReentrancyGuard, IRouter, ScheduledUpgradeable, AccessControl
     }
 
     /// @notice Relays tokens to the recipient and stores a receipt
-    /// @param token The token being relayed
+    /// @param requestId The original request ID from the source chain
+    /// @param sender The sender of the swap request on the source chain
     /// @param recipient The target recipient of the tokens
+    /// @param tokenIn The token being relayed on the destination chain
+    /// @param tokenOut The output token on the destination chain
     /// @param amountOut The amount transferred to the recipient on the destination chain
-    /// @param requestId The original request ID from the source chain
     /// @param srcChainId The ID of the source chain where the request originated
-    function relayTokens(address token, address recipient, uint256 amountOut, bytes32 requestId, uint256 srcChainId)
-        external
-        nonReentrant
-    {
+    /// @param nonce The nonce used for the swap request
+    function relayTokens(
+        bytes32 requestId,
+        address sender,
+        address recipient,
+        address tokenIn,
+        address tokenOut,
+        uint256 amountOut,
+        uint256 srcChainId,
+        uint256 nonce
+    ) external nonReentrant {
         require(!swapRequestReceipts[requestId].fulfilled, ErrorsLib.AlreadyFulfilled());
-        require(token != address(0) && recipient != address(0), ErrorsLib.InvalidTokenOrRecipient());
+        require(
+            tokenIn != address(0) && tokenOut != address(0) && sender != address(0) && recipient != address(0),
+            ErrorsLib.InvalidTokenOrRecipient()
+        );
         require(amountOut > 0, ErrorsLib.ZeroAmount());
+        require(
+            srcChainId != getChainID(),
+            ErrorsLib.SourceChainIdShouldBeDifferentFromDestination(srcChainId, getChainID())
+        );
+        require(
+            requestId
+                == keccak256(
```

Comment:
not a blocker, but I noticed building stuff earlier that forge now complains that using the `keccak256` func is less gas efficient than inline assembly for it. Tbh it seems dumb the compiler wouldn't just rewrite them, but should we swap them out?

Reply:
Using the current approach seems more readable / safer, easier to audit, and the gas difference is small to make the switch.
The higher gas savings comes from taking out the `abi.encode` and packing the fields manually but we'll need to ensure it's error free for things like addresses.
When we call `abi.encode(...)`, Solidity converts every argument into ABI-encoded words.
Each “word” is exactly 32 bytes (256 bits). Types smaller than 32 bytes like address (20 bytes) or uint8 are left-padded with zeros so they still take up a full 32-byte slot and need to be manually managed.